### ISSUE TITLE
IO.copy_stream double-writes to InternetMessageIO

### ISSUE THREAD

**Issue by kalenp** [jruby/jruby#6555]
Originally discovered when using the `Net::SMTP` library, when using `smtp_object.open_message_stream` which provides an output IO object to write to.  We were using `IO.copy_stream` to copy data to that stream, and noticed that sometimes the data would be corrupted.  

## Repro

Stripping down the original scenario, I was able to get the following reproduction script which generates some data in a StringIO and then copies it to the `Net::InternetMessageIO`, which is the class used by `Net::SMTP`.

```
require "net/protocol"
require "socket"
require "stringio"

ADDRESS = "127.0.0.1"
PORT = 3456

METHOD = ARGV[0]

TOTAL_BYTES = 32 * 1024
LINE_LENGTH = 128
LINE_COUNT = TOTAL_BYTES / LINE_LENGTH
FILLER = "a" * (LINE_LENGTH - 8)
DATA = Range.new(1, LINE_COUNT).map { |i| sprintf("% 5d %s", i, FILLER) }.join("\r\n")

StringIO.open(DATA) do |reader|
  socket = Net::InternetMessageIO.new(TCPSocket.open(ADDRESS, PORT))
  socket.write_message_by_block do |socket_writer|
    bytes_written = 0

    if METHOD == "copy_stream"
      puts "Using IO.copy_stream"
      bytes_written += IO.copy_stream(reader, socket_writer)
    elsif METHOD == "manual"
      puts "Using manual stream copying"
      while !reader.eof?
        bytes_written += socket_writer.write(reader.read(8 * 1024))
      end
    else
      raise "Unknown method: #{METHOD}"
    end

    puts "#{bytes_written} bytes written"
  end
end
```

Doing a test run, comparing against MRI
```
$ rbenv local jruby-9.2.13.0
$ 
$ netcat -l 127.0.0.1 3456 > output &
$ ruby copy_stream.rb copy_stream
Using IO.copy_stream
32766 bytes written
$ wc -c output
40961 output
$ 
$ netcat -l 127.0.0.1 3456 > output &
$ ruby copy_stream.rb manual
Using manual stream copying
32640 bytes written
$ wc -c output
32771 output
$
$ rbenv local 2.5.8
$
$ netcat -l 127.0.0.1 3456 > output &
$ ruby copy_stream.rb copy_stream
Using IO.copy_stream
32640 bytes written
$ wc -c output
32771 output
$
$ netcat -l 127.0.0.1 3456 > output &
$ ruby copy_stream.rb manual
Using manual stream copying
32640 bytes written
$ wc -c output
32771 output
$
```

You can see from the `wc -c` counts that MRI and JRuby + manual copying of blocks all output the same byte count (and looking at the files, they're all the same contents).  But JRuby + IO.copy_stream produces another ~8k worth of duplicated data.  The returned byte count is also inconsistent with JRuby + IO.copy_stream vs the others.  All of these differ in what was returned vs what was actually written, presumably because InternetMessageIO is adding some extra characters somewhere along the way (looks like an extra byte per line, plus an extra line consisting of ".\r\n").  

## Workaround

We have a workaround here of doing manual copying by blocks.

## Affected JRuby Versions

JRuby 9.2.13.0 and 9.2.14.0 have this bug
JRuby 9.2.9.0 through 9.2.12.0 seem to work correctly here

**Comment by headius** [comment]
Good find... probably similar to previous fixes for `copy_stream` when one or both sides can't be converted into a real IO channel.

**Comment by headius** [comment]
I can reproduce locally comparing CRuby 3.0 with any version of JRuby since 9.2.13.0:

```
[] ~/projects/jruby-9.2 $ rvm ruby-3.0 do ruby ../jruby/blah.rb copy_stream
Using IO.copy_stream
32640 bytes written

[] ~/projects/jruby-9.2 $ rvm jruby-9.2.13.0 do jruby ../jruby/blah.rb copy_stream
Using IO.copy_stream
32766 bytes written

[] ~/projects/jruby-9.2 $ rvm jruby-9.2.14.0 do jruby ../jruby/blah.rb copy_stream
Using IO.copy_stream
32766 bytes written

[] ~/projects/jruby-9.2 $ jruby -v ../jruby/blah.rb copy_stream
jruby 9.2.15.0-SNAPSHOT (2.5.7) 2021-02-02 73c2e10128 OpenJDK 64-Bit Server VM 11.0.4+11 on 11.0.4+11 +jit [darwin-x86_64]
Using IO.copy_stream
32766 bytes written

[] ~/projects/jruby $ jruby -v blah.rb copy_stream
jruby 9.3.0.0-SNAPSHOT (2.6.5) 2021-02-11 632ca8b10f OpenJDK 64-Bit Server VM 11.0.4+11 on 11.0.4+11 +jit [darwin-x86_64]
Using IO.copy_stream
32766 bytes written
```

Interestingly my numbers differ from yours but perhaps there were changes to the script since you posted your output. In any case, we are off by 126 bytes in what we report we wrote, and the output file has the extra 8k or so.

Looking at the output, it seems likely that we are not advancing to the next chunk, because after 255 lines are written we go to write the remaining content and it includes already-written output:

```
  255 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  256 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  193 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  194 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  195 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  196 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

**Comment by kalenp** [comment]
I'm not surprised that your numbers differ.  Even with slight changes to the script, I would get the duplicate data appearing at different points within the output.  I didn't find any obvious pattern to that sensitivity, though.

**Comment by headius** [comment]
I believe I know the problem. In my fix for writing the entire buffer, it may not be advancing the buffer position properly to track where to continue writing from, so it re-writes some content. I am trying to figure out the proper blessed way to write out all the contents of a ByteBuffer if the write call does not write everything.

**Comment by headius** [comment]
This patch gets us closer. This fixes a bug in our IO wrapping channel: it did not advance the given buffer the number of bytes it actually wrote.

Using the bytes written count, we would try to write the last chunk an additional time to finish the remaining bytes, resulting in a duplicate write.

Using `buffer.hasRemaining()` which I am using locally, the buffer would never advance and the write would keep writing repeatedly the same bytes.

However there is still some missing logic here. The final line should not show in the output because it does not have a `\r\n` in it. MRI bails out at line 255, the last line with a terminator. JRuby, with the diff above, also stops producing output at the 255th line, but then blocks waiting for the stream to accept the remaining unterminated 256th line. So it may be that we are supposed to give up on the copy_stream once we have attempted a write and no bytes are written?

**Comment by headius** [comment]
Oops, the aforementioned patch:

```diff
diff --git a/core/src/main/java/org/jruby/RubyIO.java b/core/src/main/java/org/jruby/RubyIO.java
index 381d05a831..0b87d8a75c 100644
--- a/core/src/main/java/org/jruby/RubyIO.java
+++ b/core/src/main/java/org/jruby/RubyIO.java
@@ -4453,9 +4453,8 @@ public class RubyIO extends RubyObject implements IOEncodable, Closeable, Flusha
 
             // write buffer fully and then clear
             flipBuffer(buffer);
-            long w = 0;
-            while (w < n) {
-                w += to.write(buffer);
+            while (buffer.hasRemaining()) {
+                to.write(buffer);
             }
             clearBuffer(buffer);
 
diff --git a/core/src/main/java/org/jruby/util/IOChannel.java b/core/src/main/java/org/jruby/util/IOChannel.java
index 457b07457c..7e35fe5dd1 100644
--- a/core/src/main/java/org/jruby/util/IOChannel.java
+++ b/core/src/main/java/org/jruby/util/IOChannel.java
@@ -96,15 +96,26 @@ public abstract class IOChannel implements Channel {
 
     protected int write(CallSite write, ByteBuffer src) throws IOException {
         ByteList buffer;
+        int position = src.position();
 
+        // wrap buffer contents with ByteList
         if (src.hasArray()) {
-            buffer = new ByteList(src.array(), src.position(), src.remaining(), true);
+            buffer = new ByteList(src.array(), src.position(), src.remaining(), false);
         } else {
             buffer = new ByteList(src.remaining());
             buffer.append(src, src.remaining());
         }
+
+        // call write with new String based on this ByteList
         IRubyObject written = write.call(runtime.getCurrentContext(), io, io, RubyString.newStringLight(runtime, buffer));
-        return (int)written.convertToInteger().getLongValue();
+        int wrote = written.convertToInteger().getIntValue();
+
+        // advance source buffer to match bytes written
+        if (wrote > 0) {
+            src.position(position + wrote);
+        }
+
+        return wrote;
     }
 
     protected CallSite initReadSite(String readMethod) {
```

**Comment by headius** [comment]
Looks like there is a behavioral difference for real and fake output streams in CRuby that we are not duplicating here.

CRuby appears to have the following behavior for varios types of target:

* If it is a real IO, call io_binwrite to write the output, which loops until the entire contents have been written and waits for writability if it has to write in chunks.
* If it is a fake IO, dynamically dispatch to `write` *exactly once* and use its result to increment how many bytes were written.

The fix I made in bc9a5acf5ad05384d71f4e7e0da34850b1f7c200, backported to 9.2 as 36084b52c4783ccc88168d34ce5d25b1f6840518, does the following:

* If the output is a File, use NIO FileChannel.transferFrom to consume the results, which will loop until all is written.
* For all other outputs, dynamically dispatch `write` as many times as needed to write all content.

The second behavior is correct if the output is actually an IO, but incorrect if the output is a fake IO.

In this case, it gets stuck forever trying to write the remaining bytes that the InternetMessageIO has left on the wire when instead it should "fire and forget".

I noticed this manifests as an inaccurate write count for your example:

```
[] ~/projects/jruby $ netcat -l 127.0.0.1 -p 3456 > output &
[1] 23109

[] ~/projects/jruby $ rvm ruby-3.0 do ruby ../jruby/blah.rb copy_stream
Using IO.copy_stream
32640 bytes written
[1]+  Exit 1                  netcat -l 127.0.0.1 -p 3456 > output

[] ~/projects/jruby $ wc output
     257     513   32771 output
```

This may be something to raise with CRuby, even if there's no way to fix it. The problem here is that with an opaque dynamic `write` we have to assume all bytes have been written, or at least as many bytes as the target stream wants to write have been written, and there's nothing else to be done with the remaining bytes. Either the target has consumed them, or it hasn't, but we must only report back what the target claims it actually wrote.

To fix this I will need to split up dynamic-dispatch targets from real IO targets.
